Fix Date / Datetime issues

thiagoc7 opened the pull request.

thiagoc7 committed:
minor tests refactoring

change same file names
setup_all for query tests
create trevis file

thiagoc7 committed:
convert Ecto dates to Rethink Dates

tests failing, need to provide timezone

hamiltop commented:
Does it handle dates for update as well?

hamiltop commented on lib/rethinkdb_ecto/repo.ex:
```diff
@@ -72,6 +74,30 @@ defmodule RethinkDB.Ecto.Repo do
+    {attr, value} = item
+    date_string = to_date_string(value)
+
+    cond do
```
I'm not a big `cond` fan.

I think a single pattern match would suffice here. `case value do`

hamiltop replied:
Alternatively, you could have convert_date itself pattern match on %Ecto.Date. that would be even simpler.

hamiltop commented on lib/rethinkdb_ecto/repo.ex:
```diff
@@ -72,6 +74,30 @@ defmodule RethinkDB.Ecto.Repo do
+  defp to_rethinkdb_date(value), do: RethinkDB.Query.iso8601(value)
+
+  defp to_date_string(%Ecto.Date{year: year, month: month, day: day}) do
+    %Ecto.Date{year: year, month: month, day: day}
```
If we're breaking this out already, I would look at the `time` query in RethinkDB. There's a 4 arg version and a 7 args version.

thiagoc7 committed:
refactor date conversion

now it's all in convert_date function

thiagoc7 commented:
Thanks for the code review, I feel that now it's better, but wait for your call.
For green tests I need to convert it back to Ecto.Date after find.

I'll work on this in a couple hours, but can you point me to right place?

hamiltop commented on lib/rethinkdb_ecto/repo.ex:
```diff
@@ -107,6 +111,16 @@ defmodule RethinkDB.Ecto.Repo do
     end
   end
 
+  defp convert_date({attr, %Ecto.Date{year: year, month: month, day: day}}) do
```
Love it. :smile:

thiagoc7 replied:
Thanks!!

hamiltop commented:
`load_model` is where the magic happens. Right now it's using `Ecto.Schema.__load__`. It's probably the most fragile part of the codebase, as it's a public but undocumented function in Ecto. So, "here be dragons".

There's a private function called `load` that is passed into `__load__`. Right now that function just returns `{:ok, data}`. I think if you pattern match on the RethinkDB date type, you can return `{:ok, convert_date(data)}` or something of the sort.

thiagoc7 commented:
Yeah, `load` will do!!
I'm just having a hard time to get create a `Ecto.Date%{}`.

```
iex(10)> value_from_db = %RethinkDB.Pseudotypes.Time{epoch_time: 1452643200,  timezone: "+00:00"}
iex(11)> RethinkDB.Query.day(value_from_db)
%RethinkDB.Q{query: [130,
  [%{:epoch_time => 1452643200,
     :timezone => "+00:00",
     "$reql_type$" => "TIME"}]]}
```

hamiltop commented:
http://hexdocs.pm/ecto/Ecto.Date.html#cast/1

hamiltop commented:
Hmm... cast won't do epoch time... `:calendar.seconds_to_time` is probably what you want

thiagoc7 committed:
convert back to Ecto.Date

using erlang :calendar
all green

thiagoc7 commented:
uhu, all green!!
Solution came from [this](http://michal.muskala.eu/2015/07/30/unix-timestamps-in-elixir.html) blog.
Maybe it would be nice to add this date conversion to `rethinkdb_elixir`.

hamiltop commented:
I want the driver to be pretty limited in data types. That's what libs like this are for. Some helpers to and from core datetime structures and RethinkDB structures might make sense for the driver though.

hamiltop commented on lib/rethinkdb_ecto/repo.ex:
```diff
@@ -107,6 +111,28 @@ defmodule RethinkDB.Ecto.Repo do
+    {attr, RethinkDB.Query.time(year, month, day, hour, min, sec, "Z")}
+  end
+
+  defp convert_date(%RethinkDB.Pseudotypes.Time{epoch_time: epoch_time}) do
```
Don't love this as much.

`convert_date` can accept multiple data types, but returning various data types is a bit rough. Perhaps `convert_date_to_ecto` and `convert_date_to_rethinkdb`

hamiltop replied:
Actually, even more opposed since the other `convert_date` clauses take a key value pair. This should just be a different function.

hamiltop commented:
Maybe adding the equivalent to `Ecto.Date.to_erl` and `Ecto.Date.from_erl` in the driver would be useful. What do you think?

thiagoc7 committed:
better functions names

add convert_date_from_rethinkdb, so we have a cleaner convert_date
elixir 1.2 on travis

thiagoc7 committed:
better functions names

add convert_date_from_rethinkdb, so we have a cleaner convert_date

thiagoc7 commented:
> Maybe adding the equivalent to Ecto.Date.to_erl and Ecto.Date.from_erl in the driver would be useful. > What do you think?

I think that it's that the library should deal with date, at least `to_erl`.
Maybe it can be an `attr` in the struct.

hamiltop commented:
> Maybe it can be an attr in the struct.

Not sure what you mean by that. But let's open up an issue on that repo and discuss it.

hamiltop commented on test/query_test.exs:
```diff
@@ -0,0 +1,40 @@
+    from_db = TestRepo.get(TestModel, test_model.id)
+    assert test_model == from_db
+
+    # update_changeset = TestRepo.changeset(from_db, %{title: "yayay"})
```
What's the story here with these comments?

thiagoc7 replied:
I'll create all `CRUD` tests.
The `update` assert is commented because I couldn't create a changeset, I'll take a look on this tomorrow (But if you want to give me a clue will be very welcome!).

Do you want to test the `schema`?

```
test "respect schema" do
    {:error, changeset} = TestRepo.insert(%TestModel{blablabla: "yay"})
    assert changeset.errors == "blablabla"
  end
```

It's not passing, and I don't know how to fix.

And can you authorise the repo on travis?
I copied [.travis.yml](https://github.com/hamiltop/rethinkdb-elixir/blob/master/.travis.yml) from rethink-elixir.

hamiltop replied:
Authorized on travis. I'll take a look tonight and see what's going on.

Thank you for going above and beyond on this. This repo was a quick hack one afternoon and I've been pretty blindsided by its popularity. I'll feel a lot more comfortable with people using this with these basic tests in place. It also makes it easier for people to get involved.

hamiltop commented on .travis.yml:
```diff
@@ -0,0 +1,17 @@
+language: elixir
+elixir:
+  - 1.1.1
+  - 1.2.0-rc.1
```
Let's just support 1.2 ... looks like 1.1.1 is failing on travis

thiagoc7 commented:
I'm glad to help!
I'm not a professional, code is a hobby, do this just because I like. But I don't get involved with other programmers as much. So for me it's really cool helping you this way, and I learn a lot.

I end up here because I get a little disappointed after finding out how `phoenix's channels` work... I was expecting something more `firebase`, or `meteor`.
So I looked for a solution, and `changefeeds` seems to fit.
Maybe in the future we could work on something to make phoenix realtime features as easy as firebase.

Tomorrow I'll finish the tests. And please let me know when you'll merge, so I rebase all this commits.
Thanks for the warm welcome!

Cheers!

thiagoc7 committed:
more tests

still need to fix get tests

thiagoc7 commented:
I created all `CRUD` tests, but there are two that are not working:
- `get many` - I don't know how to make db insert work in transactions, so I can't control what is in db
- `query` - where does `table` function came from? It's in readme.

hamiltop commented:
@thiagoc7 for get_many, in the past I've had a unique table per test. That gets the desired sandboxing. Or delete everything in the table before inserting.

For table, it's in RethinkDB.Query.

thiagoc7 committed:
more tests

thiagoc7 commented:
Hi @hamiltop, thanks for the tips!
Now just the `query` test is failing.
I'll let it crash on travis, so you can see the error:

```
     test/query_test.exs:46
     ** (UndefinedFunctionError) undefined function RethinkDB.Q.fetch/2
     stacktrace:
       (rethinkdb) RethinkDB.Q.fetch(%RethinkDB.Q{query: [10, [#Reference<0.0.4.503>]]}, :title)
       (elixir) lib/access.ex:77: Access.get/3
       test/query_test.exs:51: anonymous fn/1 in QueryTest.test filtered queries work/1
       (rethinkdb) lib/rethinkdb/query.ex:1808: RethinkDB.Query.func/1
       (rethinkdb) lib/rethinkdb/query.ex:1035: RethinkDB.Query.filter/2
       test/query_test.exs:50
```

hamiltop commented:
Ahh... My example was flawed. You need to use the lambda macro from
RethinkDB.Lambda like this:

import RethinkDB.Lambda

filter(lambda fn post ->
  post[:title] == "yayay"
end)

Could you update the readme too?

On Thu, Jan 14, 2016, 8:35 AM Thiago notifications@github.com wrote:

> Hi @hamiltop https://github.com/hamiltop, thanks for the tips!
> Now just the query test is failing.
> I'll let it crash on travis, so you can see the error:
> 
> ```
>  test/query_test.exs:46
>  ** (UndefinedFunctionError) undefined function RethinkDB.Q.fetch/2
>  stacktrace:
>    (rethinkdb) RethinkDB.Q.fetch(%RethinkDB.Q{query: [10, [#Reference<0.0.4.503>]]}, :title)
>    (elixir) lib/access.ex:77: Access.get/3
>    test/query_test.exs:51: anonymous fn/1 in QueryTest.test filtered queries work/1
>    (rethinkdb) lib/rethinkdb/query.ex:1808: RethinkDB.Query.func/1
>    (rethinkdb) lib/rethinkdb/query.ex:1035: RethinkDB.Query.filter/2
>    test/query_test.exs:50
> ```
> 
> —
> Reply to this email directly or view it on GitHub
> https://github.com/hamiltop/rethinkdb_ecto/pull/9#issuecomment-171692906
> .

hamiltop commented:
Actually, I'm going to redo the readme anyway, I can take care of it.

On Thu, Jan 14, 2016, 8:59 AM Peter Hamilton peterghamilton@gmail.com
wrote:

> Ahh... My example was flawed. You need to use the lambda macro from
> RethinkDB.Lambda like this:
> 
> import RethinkDB.Lambda
> 
> filter(lambda fn post ->
>   post[:title] == "yayay"
> end)
> 
> Could you update the readme too?
> 
> On Thu, Jan 14, 2016, 8:35 AM Thiago notifications@github.com wrote:
> 
> > Hi @hamiltop https://github.com/hamiltop, thanks for the tips!
> > Now just the query test is failing.
> > I'll let it crash on travis, so you can see the error:
> > 
> > ```
> >  test/query_test.exs:46
> >  ** (UndefinedFunctionError) undefined function RethinkDB.Q.fetch/2
> >  stacktrace:
> >    (rethinkdb) RethinkDB.Q.fetch(%RethinkDB.Q{query: [10, [#Reference<0.0.4.503>]]}, :title)
> >    (elixir) lib/access.ex:77: Access.get/3
> >    test/query_test.exs:51: anonymous fn/1 in QueryTest.test filtered queries work/1
> >    (rethinkdb) lib/rethinkdb/query.ex:1808: RethinkDB.Query.func/1
> >    (rethinkdb) lib/rethinkdb/query.ex:1035: RethinkDB.Query.filter/2
> >    test/query_test.exs:50
> > ```
> > 
> > —
> > Reply to this email directly or view it on GitHub
> > https://github.com/hamiltop/rethinkdb_ecto/pull/9#issuecomment-171692906
> > .

thiagoc7 committed:
all tests green

thiagoc7 commented:
now it's all green!

I noticed that you are not validating before insert.
the date test's shouldn't work, because I set `title` as a required field on changesets

hamiltop commented:
So I think this is good to go.  I'll go ahead and merge and close this.

Re: validating schema. Let's have a discussion on https://github.com/hamiltop/rethinkdb_ecto/issues/10 around that.

Thanks for the help!
